nbs/05_preprocessing/py/05-gtex_v8-select_top_genes.py

```diff
@@ -0,0 +1,519 @@
+# ---
+# jupyter:
+#   jupytext:
+#     cell_metadata_filter: all,-execution,-papermill,-trusted
+#     text_representation:
+#       extension: .py
+#       format_name: percent
+#       format_version: '1.3'
+#       jupytext_version: 1.11.4
+#   kernelspec:
+#     display_name: Python 3 (ipykernel)
+#     language: python
+#     name: python3
+# ---
+
+# %% [markdown] tags=[]
+# # Description
+
+# %% [markdown] tags=[]
+# It analyzes different strategies to take the genes from GTEx data with the highest variability, being this variability measured with different strategies: variance (`var`), coefficient of variation (`cv`) and mean absolute variation (`mad`) applied on two different versions of the data: 1) the raw TPM-normalized gene expression data (here refered to as `raw`), and 2) the log2-transformed version of the raw data (here refered to as `log2`).
+
+# %% [markdown] tags=[]
+# # Modules
+
+# %% tags=[]
+import numpy as np
+from scipy.spatial.distance import pdist, squareform
+import pandas as pd
+import matplotlib.pyplot as plt
+import seaborn as sns
+from sklearn.decomposition import PCA
+from sklearn.preprocessing import scale
+from tqdm import tqdm
+
+from clustermatch import conf
+
+# %% [markdown] tags=[]
+# # Settings
+
+# %% tags=[]
+N_TOP_GENES_MAX_VARIANCE = 5000
+
+# %% [markdown] tags=[]
+# # Paths
+
+# %% tags=[]
+INPUT_DIR = conf.GTEX["DATA_DIR"] / "data_by_tissue"
+display(INPUT_DIR)
+
+# %% tags=[]
+OUTPUT_DIR = conf.GTEX["GENE_SELECTION_DIR"]
+OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
+display(OUTPUT_DIR)
+
+# %% [markdown] tags=[]
+# # Functions
+
+# %% tags=[]
+PCA_OPTIONS = {
+    "n_components": 5,
+    "random_state": 0,
+}
+
+
+# %% tags=[]
+def standardize(data):
+    return pd.DataFrame(
+        data=scale(data),
+        index=data.index.copy(),
+        columns=data.columns.copy(),
+    )
+
+
+# %% tags=[]
+def plot_pca(data, std=True):
+    if std:
+        data = standardize(data)
+
+    clf = PCA(**PCA_OPTIONS)
+    pca_data = clf.fit_transform(data)
+
+    pca_data = pd.DataFrame(
+        data=pca_data,
+        index=data.index.copy(),
+        columns=[f"PCA{i+1}" for i in range(pca_data.shape[1])],
+    )
+
+    g = sns.pairplot(data=pca_data)
+    display(g)
+
+
+# %% [markdown] tags=[]
+# # Compare different criteria to select highly variable genes
+
+# %% tags=[]
+# I will store here the top genes selected by each method
+top_genes_var = {}
+
+# %% [markdown] tags=[]
+# ## Get test data
+
+# %% tags=[]
+test_data = pd.read_pickle(INPUT_DIR / "gtex_v8_data_whole_blood.pkl")
+
+# %% tags=[]
+test_data.shape
+
+# %% tags=[]
+test_data.head()
+
+# %% [markdown] tags=[]
+# ## Get test data in log2
+
+# %% tags=[]
+log2_test_data = np.log2(test_data)
+
+# %% tags=[]
+log2_test_data.head()
+
+
+# %% tags=[]
+def replace_by_minimum(sample_data):
```

Comment:
> I don't know if this will make a big difference, but is there a reason you didn't set these negative values to 0 instead of the minimum?

That was actually my first attempt. But then I was reading what others have done and I changed my mind. Let me rethink that here: so a log2 value of `-inf` is a zero in the raw expression data (TPM-normalized). A negative log2 value represents a raw expression less than 1, so I think it's not correct to replace negative log2 values by 0, right? They can actually be anything from 0 to 1 (in the raw data).

But after thinking this again, you are right in that I should put 0 in all `-inf` entries. A 0 in a sample should not be differentiated from a 0 in another sample, right? No-expression is no-expression in any sample. I think that makes sense and I will change it. However, ...

> You could also try adding a pseudocount so `np.log2(test_data + 1)`. The amount you add will depend on the log that you're taking.

... this is a great solution, awesome! So by adding 1 to the raw expression data, the logarithm is always defined, a log2 value of 0 is a 0 in the raw expression. Awesome, thank you, Alex!